File: Napakalaki.ruby/CardDealer.rb
#encoding: utf-8
require_relative 'Monster'
require_relative 'Treasure'
require_relative 'Cultist'
require_relative 'BadConsequence'
require_relative 'NumericBadConsequence'
require_relative 'SpecificBadConsequence'
require_relative 'DeathBadConsequence'
require 'singleton'

module NapakalakiGame

class CardDealer
  include Singleton

  def initialize
    @unusedTreasures = []
    @usedTreasures = []
    @unusedMonsters = []
    @usedMonsters = []
    @unusedCultists = []
  end

private
  def initTreasureCardDeck

      @unusedTreasures = []

      @unusedTreasures << Treasure.new("¡Sí mi amo!", 4, TreasureKind::HELMET)
      @unusedTreasures << Treasure.new("Botas de investigación", 3, TreasureKind::SHOES)
      @unusedTreasures << Treasure.new("Capucha de Cthulhu", 3, TreasureKind::HELMET)
      @unusedTreasures << Treasure.new("A prueba de babas", 2, TreasureKind::ARMOR)
      @unusedTreasures << Treasure.new("Botas de lluvia  ́acida", 1, TreasureKind::BOTHHANDS)
      @unusedTreasures << Treasure.new("Casco minero", 2, TreasureKind::HELMET)
      @unusedTreasures << Treasure.new("Ametralladora Thompson", 4, TreasureKind::BOTHHANDS)
      @unusedTreasures << Treasure.new("Camiseta de la UGR", 1, TreasureKind::ARMOR)
      @unusedTreasures << Treasure.new("Clavo de rail ferroviario", 3, TreasureKind::ONEHAND)
      @unusedTreasures << Treasure.new("Cuchillo de sushi arcano", 2, TreasureKind::ONEHAND)
      @unusedTreasures << Treasure.new("Fez al ́opodo", 3, TreasureKind::HELMET)
      @unusedTreasures << Treasure.new("Hacha prehistóorica", 2, TreasureKind::ONEHAND)
      @unusedTreasures << Treasure.new("El aparato del Pr. Tesla", 4, TreasureKind::ARMOR)
      @unusedTreasures << Treasure.new("Gaita", 4, TreasureKind::BOTHHANDS)
      @unusedTreasures << Treasure.new("Insecticida", 2, TreasureKind::ONEHAND)
      @unusedTreasures << Treasure.new("Escopeta de 3 cañones", 4, TreasureKind::BOTHHANDS)
      @unusedTreasures << Treasure.new("Garabato místico", 2, TreasureKind::ONEHAND)
      @unusedTreasures << Treasure.new("La rebeca metálica", 2, TreasureKind::ARMOR)
      @unusedTreasures << Treasure.new("Lanzallamas", 4, TreasureKind::BOTHHANDS)
      @unusedTreasures << Treasure.new("Necro-comicón", 1, TreasureKind::ONEHAND)
      @unusedTreasures << Treasure.new("Necronomicón", 5, TreasureKind::BOTHHANDS)
      @unusedTreasures << Treasure.new("Linterna a 2 manos", 3, TreasureKind::BOTHHANDS)
      @unusedTreasures << Treasure.new("Necro-gnomicón", 2, TreasureKind::ONEHAND)
      @unusedTreasures << Treasure.new("Necrotelecom", 2, TreasureKind::HELMET)
      @unusedTreasures << Treasure.new("Mazo de los antiguos", 3, TreasureKind::ONEHAND)
      @unusedTreasures << Treasure.new("Necro-playboycón", 3, TreasureKind::ONEHAND)
      @unusedTreasures << Treasure.new("Porra preternatural", 2, TreasureKind::ONEHAND)
      @unusedTreasures << Treasure.new("Shogulador", 1, TreasureKind::BOTHHANDS)
      @unusedTreasures << Treasure.new("Varita de atizamiento", 3, TreasureKind::ONEHAND)
      @unusedTreasures << Treasure.new("Tentáculo de pega", 2, TreasureKind::HELMET)
      @unusedTreasures << Treasure.new("Zapato deja-amigos", 1, TreasureKind::SHOES)

  end

  def initMonsterCardDeck

    @unusedMonsters = []

    badC = SpecificBadConsequence.new('Pierdes tu armadura visible y otra oculta.', 0,
    [TreasureKind::ARMOR], [TreasureKind::ARMOR])
    premio = Prize.new(2, 1)
    @unusedMonsters << Monster.new('3 Byakhees de bonanza', 8, premio, badC, 0)

    badC = SpecificBadConsequence.new('Embobados con el lindo primigenio te descartas de tu casco visible.', 0,
    [TreasureKind::HELMET], 0)
    premio = Prize.new(1, 1)
    @unusedMonsters << Monster.new('Chibithulhu', 2, premio, badC, 0)

    badC = SpecificBadConsequence.new('El primordial bostezo contagioso. Pierdes el calzado visible.', 0,
    [TreasureKind::SHOES], 0)
    premio = Prize.new(1, 1)
    @unusedMonsters << Monster.new('El sopor de Dunwich', 2, premio, badC, 0)

    badC = SpecificBadConsequence.new('Te atrapan para llevarte de fiesta y te dejan caer en mitad del vuelo. Descarta 1 mano visible y 1 mano oculta.',
     0, [TreasureKind::ONEHAND], [TreasureKind::ONEHAND])
    premio = Prize.new(4, 1)
    @unusedMonsters << Monster.new('Angeles de la noche ibicenca', 14, premio, badC, 0)

    badC = NumericBadConsequence.new('Pierdes todos tus tesoros visibles.', 0, 10, 0)
    premio = Prize.new(3, 1)
    @unusedMonsters << Monster.new('El gorron en el umbral', 10, premio, badC, 0)

    badC = SpecificBadConsequence.new('Pierdes la armadura visible.', 0,  [TreasureKind::ARMOR], 0)
    premio = Prize.new(2, 1)
    @unusedMonsters << Monster.new('H.P. Munchcraft', 6, premio, badC, 0)

    badC = SpecificBadConsequence.new('Sientes bichos bajo la ropa. Descarta la armadura visible.', 0,
    [TreasureKind::ARMOR], 0)
    premio = Prize.new(1, 1)
    @unusedMonsters << Monster.new('Bichgooth', 2, premio, badC, 0)

    badC = NumericBadConsequence.new('Pierdes 5 niveles y 3 tesoros visibles.', 5, 3, 0)
    premio = Prize.new(4, 2)
    @unusedMonsters << Monster.new('El rey de rosa', 13, premio, badC, 0)

    badC = NumericBadConsequence.new('Toses los pulmones y pierdes 2 niveles.', 2, 0, 0)
    premio = Prize.new(1, 1)
    @unusedMonsters << Monster.new('La que redacta en las tinieblas', 2, premio, badC, 0)

    badC = DeathBadConsequence.new('Estos monstruos resultan bastante superficiales y te aburren mortalmente. Estas muerto.')
    premio = Prize.new(2, 1)
    @unusedMonsters << Monster.new('Los hondos', 8, premio, badC, 0)

    badC = NumericBadConsequence.new('Pierdes 2 niveles y 2 tesoros ocultos.', 2, 0, 2)
    premio = Prize.new(2, 1)
    @unusedMonsters << Monster.new('Semillas Cthulhu', 4, premio, badC, 0)

    badC = SpecificBadConsequence.new('Te intentas escaquear. Pierdes una mano visible.', 0,  [TreasureKind::ONEHAND], 0)
    premio = Prize.new(2, 1)
    @unusedMonsters << Monster.new('Dameargo', 1, premio, badC, 0)

    badC = NumericBadConsequence.new('Da mucho asquito. Pierdes 3 niveles.', 3, 0, 0)
    premio = Prize.new(1, 1)
    @unusedMonsters << Monster.new('Pollipolipo volante', 3, premio, badC, 0)

    badC = DeathBadConsequence.new('No le hace gracia que pronuncien mal su nombre. Estas muerto.')
    premio = Prize.new(3, 1)
    @unusedMonsters << Monster.new('Yskhtihyssg-Goth', 12, premio, badC, 0)

    badC = DeathBadConsequence.new('La familia te atrapa. Est́as muerto.')
    premio = Prize.new(4, 1)
    @unusedMonsters << Monster.new('Familia feliz', 1, premio, badC, 0)

    badC = SpecificBadConsequence.new('La quinta directiva primaria te obliga a perder 2 niveles y un tesoro 2 manos visible.',
    2, [TreasureKind::BOTHHANDS], 0)
    premio = Prize.new(2, 1)
    @unusedMonsters << Monster.new('Roboggoth', 8, premio, badC, 0)

    badC = SpecificBadConsequence.new('Te asusta en la noche. Pierdes un casco visible.', 0,
    [TreasureKind::HELMET], 0)
    premio = Prize.new(1, 1)
    @unusedMonsters << Monster.new('El espia', 5, premio, badC, 0)

    badC = NumericBadConsequence.new('Menudo susto te llevas. Pierdes 2 niveles y 5 tesoros visibles.', 2, 5, 0)
    premio = Prize.new(1, 1)
    @unusedMonsters << Monster.new('El Lenguas', 20, premio, badC, 0)

    badC = SpecificBadConsequence.new('Te faltan manos para tanta cabeza. Pierdes 3 niveles y tus tesoros visibles de las manos.',
    3, [TreasureKind::ONEHAND, TreasureKind::ONEHAND, TreasureKind::BOTHHANDS], 0)
    premio = Prize.new(1, 1)
    @unusedMonsters << Monster.new('Bicefalo', 20, premio, badC, 0)



    badC = SpecificBadConsequence.new("Pierdes 1 mano visible.", 0, [TreasureKind::ONEHAND], 0)
    premio = Prize.new(3, 1)
    @unusedMonsters << Monster.new("El mal indecible impronunciable", 10, badC, premio, -2)

    badC = NumericBadConsequence.new("Pierdes tus tesoros visibles. Jajaja.", 0, 10, 0)
    premio = Prize.new(2, 1)
    @unusedMonsters << Monster.new("Testigos Oculares", 6, badC, premio, 2)

    badC = DeathBadConsequence.new("Hoy no es tu día de suerte. Mueres.")
    premio = Prize.new(2, 5)
    @unusedMonsters << Monster.new("El gran cthulhu", 20, badC, premio, 4)

    badC = NumericBadConsequence.new("Tu gobierno te recorta 2 niveles.", 2, 0, 0)
    premio = Prize.new(2, 1)
    @unusedMonsters << Monster.new("Serpiente Político", 8, badC, premio, -2)

    badC = SpecificBadConsequence.new("Pierdes tu casco y tu armadura visible. Pierdes tus manos ocultas.", 0, [TreasureKind::HELMET, TreasureKind::ARMOR], [TreasureKind::ONEHAND, TreasureKind::ONEHAND, TreasureKind::ONEHAND, TreasureKind::ONEHAND, TreasureKind::BOTHHANDS, TreasureKind::BOTHHANDS, TreasureKind::BOTHHANDS, TreasureKind::BOTHHANDS])
    premio = Prize.new(1, 1)
    @unusedMonsters << Monster.new("Felpuggoth", 2, badC, premio, 5)

    badC = NumericBadConsequence.new("Pierdes 2 niveles.", 2, 0, 0)
    premio = Prize.new(4, 2)
    @unusedMonsters << Monster.new("Shoggoth", 16, badC, premio, -4)

    badC = NumericBadConsequence.new("Pintalabios negro. Pierdes 2 niveles.", 2, 0, 0)
    premio = Prize.new(1, 1)
    @unusedMonsters << Monster.new("Lolitagooth", 2, badC, premio, 3)

  end

  def initCultistsCardDeck
    @nusedCultists = []

    @unusedCultists << Cultist.new("Agaricus", 1)
    @unusedCultists << Cultist.new("Boletus", 2)
    @unusedCultists << Cultist.new("Daldinia", 1)
    @unusedCultists << Cultist.new("Bolbitius", 2)
    @unusedCultists << Cultist.new("Calvatia", 1)
    @unusedCultists << Cultist.new("Dermoloma", 1)
  end

  def shuffleTreasures
      @unusedTreasures.shuffle!
  end

  def shuffleMonsters
      @unusedMonsters.shuffle!
  end

  def shuffleCultists
      @unusedCultists.shuffle!
  end

public

  def GetInstance
    CardDealer.instance
  end

  def nextTreasure
    if @unusedTreasures.empty?
          @unusedTreasures = @usedTreasures
          shuffleTreasures
    end
    tesoro = @unusedTreasures.at(0)
    @unusedTreasures.delete_at(0)
    giveTreasureBack(tesoro)
    tesoro
  end

  def nextMonster
    if @unusedMonsters.empty?
        @unusedMonsters = @usedMonsters
        shuffleTreasures
    end
    monstruo = @unusedMonsters.at(0)
    @unusedMonsters.delete_at(0)
    giveTreasureBack(monstruo)
    monstruo
  end

  def nextCultist
      @unusedCultists.delete_at(0)
  end

  def giveTreasureBack(t)
      @usedTreasures << t
  end

  def giveMonsterBack(m)
      @usedMonsters << m
  end

  def initCards
    initMonsterCardDeck
    shuffleMonsters
    initTreasureCardDeck
    shuffleTreasures
    initCultistsCardDeck
    shuffleCultists
  end
end

end

File: Napakalaki.java/src/NapakalakiGame/CardDealer.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;

/**
 *
 * @author victor
 */
public class CardDealer {

    private static CardDealer instance = null;
    private ArrayList<Monster> unusedMonsters;
    private ArrayList<Monster> usedMonsters;
    private ArrayList<Treasure> unusedTreasures;
    private ArrayList<Treasure> usedTreasures;
    private ArrayList<Cultist> unusedCultists;
 

  private CardDealer() { 
      unusedMonsters = null;
      usedMonsters = new ArrayList();
      unusedTreasures = null;
      usedTreasures = new ArrayList();
      unusedCultists = null;
  }
 

  private void initTreasureCardDeck(){
      
      unusedTreasures = new ArrayList();
      
      unusedTreasures.add(new Treasure("¡Sí mi amo!", 4, TreasureKind.HELMET));
      unusedTreasures.add(new Treasure("Botas de investigación", 3, TreasureKind.SHOES));
      unusedTreasures.add(new Treasure("Capucha de Cthulhu", 3, TreasureKind.HELMET));
      unusedTreasures.add(new Treasure("A prueba de babas", 2, TreasureKind.ARMOR));
      unusedTreasures.add(new Treasure("Botas de lluvia  ́acida", 1, TreasureKind.BOTHHANDS));
      unusedTreasures.add(new Treasure("Casco minero", 2, TreasureKind.HELMET));
      unusedTreasures.add(new Treasure("Ametralladora Thompson", 4, TreasureKind.BOTHHANDS));
      unusedTreasures.add(new Treasure("Camiseta de la UGR", 1, TreasureKind.ARMOR));
      unusedTreasures.add(new Treasure("Clavo de rail ferroviario", 3, TreasureKind.ONEHAND));
      unusedTreasures.add(new Treasure("Cuchillo de sushi arcano", 2, TreasureKind.ONEHAND));
      unusedTreasures.add(new Treasure("Fez al ́opodo", 3, TreasureKind.HELMET));
      unusedTreasures.add(new Treasure("Hacha prehistóorica", 2, TreasureKind.ONEHAND));
      unusedTreasures.add(new Treasure("El aparato del Pr. Tesla", 4, TreasureKind.ARMOR));
      unusedTreasures.add(new Treasure("Gaita", 4, TreasureKind.BOTHHANDS));
      unusedTreasures.add(new Treasure("Insecticida", 2, TreasureKind.ONEHAND));
      unusedTreasures.add(new Treasure("Escopeta de 3 cañones", 4, TreasureKind.BOTHHANDS));
      unusedTreasures.add(new Treasure("Garabato místico", 2, TreasureKind.ONEHAND));
      unusedTreasures.add(new Treasure("La rebeca metálica", 2, TreasureKind.ARMOR));
      unusedTreasures.add(new Treasure("Lanzallamas", 4, TreasureKind.BOTHHANDS));
      unusedTreasures.add(new Treasure("Necro-comicón", 1, TreasureKind.ONEHAND));
      unusedTreasures.add(new Treasure("Necronomicón", 5, TreasureKind.BOTHHANDS));
      unusedTreasures.add(new Treasure("Linterna a 2 manos", 3, TreasureKind.BOTHHANDS));
      unusedTreasures.add(new Treasure("Necro-gnomicón", 2, TreasureKind.ONEHAND));
      unusedTreasures.add(new Treasure("Necrotelecom", 2, TreasureKind.HELMET));
      unusedTreasures.add(new Treasure("Mazo de los antiguos", 3, TreasureKind.ONEHAND));
      unusedTreasures.add(new Treasure("Necro-playboycón", 3, TreasureKind.ONEHAND));
      unusedTreasures.add(new Treasure("Porra preternatural", 2, TreasureKind.ONEHAND));
      unusedTreasures.add(new Treasure("Shogulador", 1, TreasureKind.BOTHHANDS));
      unusedTreasures.add(new Treasure("Varita de atizamiento", 3, TreasureKind.ONEHAND));
      unusedTreasures.add(new Treasure("Tentáculo de pega", 2, TreasureKind.HELMET));
      unusedTreasures.add(new Treasure("Zapato deja-amigos", 1, TreasureKind.SHOES));
      
      
  }

  private void initMonsterCardDeck(){
      
        BadConsequence Bc;
        Prize P;
        unusedMonsters = new ArrayList();
        
        Bc = new SpecificBadConsequence("Pierdes tu armadura visible y otra oculta.", 0, new ArrayList(Arrays.asList(TreasureKind.ARMOR)), new ArrayList(Arrays.asList(TreasureKind.ARMOR)));
        P = new Prize(2, 1);
        unusedMonsters.add(new Monster("3 Byakhees de bonanza", 8, Bc, P));
        
        Bc = new SpecificBadConsequence("Embobados con el lindo primigenio te descartas de tu casco visible.", 0, new ArrayList(Arrays.asList(TreasureKind.HELMET)), new ArrayList<TreasureKind>());
        P = new Prize(1, 1);
        unusedMonsters.add(new Monster("Chibithulhu", 2, Bc, P));
        
        Bc = new SpecificBadConsequence("El primordial bostezo contagioso. Pierdes el calzado visible.", 0, new ArrayList(Arrays.asList(TreasureKind.SHOES)), new ArrayList<TreasureKind>());
        P = new Prize(1, 1);
        unusedMonsters.add(new Monster("El sopor de Dunwich", 2, Bc, P));
        
        Bc = new SpecificBadConsequence("Te atrapan para llevarte de fiesta y te dejan caer en mitad del vuelo. Descarta 1 mano visible y 1 mano oculta.", 0, new ArrayList(Arrays.asList(TreasureKind.ONEHAND)), new ArrayList(Arrays.asList(TreasureKind.ONEHAND)));
        P = new Prize(4, 1);
        unusedMonsters.add(new Monster("Angeles de la noche ibicenca", 14, Bc, P));
        
        Bc = new NumericBadConsequence("Pierdes todos tus tesoros visibles.", 0, 10, 0);
        P = new Prize(3, 1);
        unusedMonsters.add(new Monster("El gorron en el umbral", 10, Bc, P));
        
        Bc = new SpecificBadConsequence("Pierdes la armadura visible.", 0, new ArrayList(Arrays.asList(TreasureKind.ARMOR)), new ArrayList<TreasureKind>());
        P = new Prize(2, 1);
        unusedMonsters.add(new Monster("H.P. Munchcraft", 6, Bc, P));
        
        Bc = new SpecificBadConsequence("Sientes bichos bajo la ropa. Descarta la armadura visible.", 0, new ArrayList(Arrays.asList(TreasureKind.ARMOR)), new ArrayList<TreasureKind>());
        P = new Prize(1, 1);
        unusedMonsters.add(new Monster("Bichgooth", 2, Bc, P));
        
        Bc = new NumericBadConsequence("Pierdes 5 niveles y 3 tesoros visibles.", 5, 3, 0);
        P = new Prize(4, 2);
        unusedMonsters.add(new Monster("El rey de rosa", 13, Bc, P));
        
        Bc = new NumericBadConsequence("Toses los pulmones y pierdes 2 niveles.", 2, 0, 0);
        P = new Prize(1, 1);
        unusedMonsters.add(new Monster("La que redacta en las tinieblas", 2, Bc, P));
        
        Bc = new DeathBadConsequence("Estos monstruos resultan bastante superficiales y te aburren mortalmente. Estas muerto.");
        P = new Prize(2, 1);
        unusedMonsters.add(new Monster("Los hondos", 8, Bc, P));
        
        Bc = new NumericBadConsequence("Pierdes 2 niveles y 2 tesoros ocultos.", 2, 0, 2);
        P = new Prize(2, 1);
        unusedMonsters.add(new Monster("Semillas Cthulhu", 4, Bc, P));
        
        Bc = new SpecificBadConsequence("Te intentas escaquear. Pierdes una mano visible.", 0, new ArrayList(Arrays.asList(TreasureKind.ONEHAND)), new ArrayList<TreasureKind>());
        P = new Prize(2, 1);
        unusedMonsters.add(new Monster("Dameargo", 1, Bc, P));
        
        Bc = new NumericBadConsequence("Da mucho asquito. Pierdes 3 niveles.", 3, 0, 0);
        P = new Prize(1, 1);
        unusedMonsters.add(new Monster("Pollipolipo volante", 3, Bc, P));
        
        Bc = new DeathBadConsequence("No le hace gracia que pronuncien mal su nombre. Estas muerto.");
        P = new Prize(3, 1);
        unusedMonsters.add(new Monster("Yskhtihyssg-Goth", 12, Bc, P));
        
        Bc = new DeathBadConsequence("La familia te atrapa. Est́as muerto.");
        P = new Prize(4, 1);
        unusedMonsters.add(new Monster("Familia feliz", 1, Bc, P));
        
        Bc = new SpecificBadConsequence("La quinta directiva primaria te obliga a perder 2 niveles y un tesoro 2 manos visible.", 2, new ArrayList(Arrays.asList(TreasureKind.BOTHHANDS)), new ArrayList<TreasureKind>());
        P = new Prize(2, 1);
        unusedMonsters.add(new Monster("Roboggoth", 8, Bc, P));
        
        Bc = new SpecificBadConsequence("Te asusta en la noche. Pierdes un casco visible.", 0, new ArrayList(Arrays.asList(TreasureKind.HELMET)), new ArrayList<TreasureKind>());
        P = new Prize(1, 1);
        unusedMonsters.add(new Monster("El espia", 5, Bc, P));
        
        Bc = new NumericBadConsequence("Menudo susto te llevas. Pierdes 2 niveles y 5 tesoros visibles.", 2, 5, 0);
        P = new Prize(1, 1);
        unusedMonsters.add(new Monster("El Lenguas", 20, Bc, P));
        
        Bc = new SpecificBadConsequence("Te faltan manos para tanta cabeza. Pierdes 3 niveles y tus tesoros visibles de las manos.", 3, new ArrayList(Arrays.asList(TreasureKind.ONEHAND, TreasureKind.ONEHAND, TreasureKind.BOTHHANDS)), new ArrayList<TreasureKind>());
        P = new Prize(1, 1);
        unusedMonsters.add(new Monster("Bicefalo", 20, Bc, P));
        
        Bc = new SpecificBadConsequence("Pierdes 1 mano visible.", 0, new ArrayList(Arrays.asList(TreasureKind.ONEHAND)), new ArrayList<TreasureKind>());
        P = new Prize(3, 1);
        unusedMonsters.add(new Monster("El mal indecible impronunciable", 10, Bc, P, -2));
        
        Bc = new NumericBadConsequence("Pierdes tus tesoros visibles. Jajaja.", 0, 10, 0);
        P = new Prize(2, 1);
        unusedMonsters.add(new Monster("Testigos Oculares", 6, Bc, P, 2));
        
        Bc = new DeathBadConsequence("Hoy no es tu día de suerte. Mueres.");
        P = new Prize(2, 5);
        unusedMonsters.add(new Monster("El gran cthulhu", 20, Bc, P, 4));
        
        Bc = new NumericBadConsequence("Tu gobierno te recorta 2 niveles.", 2, 0, 0);
        P = new Prize(2, 1);
        unusedMonsters.add(new Monster("Serpiente Político", 8, Bc, P, -2));   
        
        Bc = new SpecificBadConsequence("Pierdes tu casco y tu armadura visible. Pierdes tus manos ocultas.", 0, new ArrayList(Arrays.asList(TreasureKind.HELMET, TreasureKind.ARMOR)),new ArrayList(Arrays.asList(TreasureKind.ONEHAND, TreasureKind.ONEHAND, TreasureKind.ONEHAND, TreasureKind.ONEHAND, TreasureKind.BOTHHANDS, TreasureKind.BOTHHANDS, TreasureKind.BOTHHANDS, TreasureKind.BOTHHANDS)));
        P = new Prize(1, 1);
        unusedMonsters.add(new Monster("Felpuggoth", 2, Bc, P, 5));
        
        Bc = new NumericBadConsequence("Pierdes 2 niveles.", 2, 0, 0);
        P = new Prize(4, 2);
        unusedMonsters.add(new Monster("Shoggoth", 16, Bc, P, -4));
        
        Bc = new NumericBadConsequence("Pintalabios negro. Pierdes 2 niveles.", 2, 0, 0);
        P = new Prize(1, 1);
        unusedMonsters.add(new Monster("Lolitagooth", 2, Bc, P, 3)); 
  }
  
  private void initCultistsCardDeck() {
      
      unusedCultists = new ArrayList();
      
      unusedCultists.add(new Cultist("Agaricus", 1));
      unusedCultists.add(new Cultist("Boletus", 2));
      unusedCultists.add(new Cultist("Daldinia", 1));
      unusedCultists.add(new Cultist("Bolbitius", 2));
      unusedCultists.add(new Cultist("Calvatia", 1));
      unusedCultists.add(new Cultist("Dermoloma", 1));
  }
  
  private void shuffleTreasures(){
      Collections.shuffle(unusedTreasures);
  }

  private void shuffleMonsters(){
      Collections.shuffle(unusedMonsters);
  }
  
  private void shuffleCultists () {
      Collections.shuffle(unusedCultists);
  }
  
  public static CardDealer getInstance() {
      if (instance == null)
          instance = new CardDealer();
      return instance;
  }
  
  public Treasure nextTreasure(){
      if (unusedTreasures == null){
          unusedTreasures = usedTreasures;
          shuffleTreasures();
        }  
      Treasure tesoro = unusedTreasures.remove(0);
      giveTreasureBack(tesoro);
      return tesoro;
  }

  public Monster nextMonster(){
      if (unusedMonsters == null){
          unusedMonsters = usedMonsters;
          shuffleMonsters();
        }
      Monster monstruo = unusedMonsters.remove(0);
      giveMonsterBack(monstruo);
      return monstruo;
  }
  
  public Cultist nextCultist () {
      Cultist sectario = unusedCultists.remove(0);
      return sectario;
  }

  public void giveTreasureBack(Treasure t){
      usedTreasures.add(t);
  }

  public void giveMonsterBack(Monster m){
      usedMonsters.add(m);
  }

  public void initCards(){
      initMonsterCardDeck();
      shuffleMonsters();
      initTreasureCardDeck();
      shuffleTreasures();
      initCultistsCardDeck();
      shuffleCultists ();
  }
}
